specs/interop.md

```diff
@@ -0,0 +1,324 @@
+| `MAX_LOG_DATA_SIZE` | `uint64(2**24)` or 16,777,216 |
+| `MAX_MESSAGES_PER_BLOCK` | `uint64(2**20)` or 1,048,576 |
+
+```python
+class CrossChainMessage(Container):
+    nonce: uint256
+    sender: ExecutionAddress
+    target: ExecutionAddress
+    value: uint256
+    timestamp: uint256
+    chainid: uint256
+    data: ByteList[MAX_LOG_DATA_SIZE]
+
+messages = List[CrossChainMessage, MAX_MESSAGES_PER_BLOCK]
+
+block.extra_data = messages.hash_tree_root()
+```
+
+## Predeploys
+
+Two new system level predeploys are introduced for managing cross chain messaging.
+
+### CrossL2Outbox
+
+Address: `0x4200000000000000000000000000000000000023`
+
+The `CrossL2Outbox` is responsible for initiating a cross chain message on the local chain.
+For every cross chain message that is sent, it MUST emit an event that indicates the existence
+of the cross chain message.
+
+The event SHOULD make it as cheap as possible for the remote chain to look up the existence
+of the initiating transaction when verifying the relaying transaction.
+
+#### Initiating Messages
+
+An initiating message emits the following event:
+
+```solidity
+event MessagePassed(
+    bytes32 indexed messageHash,
+    address indexed sender,
+    address indexed target,
+    uint256 nonce,
+    uint256 value,
+    uint256 timestamp,
+    uint256 chainid,
+    bytes data
+);
+```
+
+The `messageHash` and chain id uniquely identify the existence of an initiating message and SHOULD
+be used by a block builder to ensure the existence of the initiating message when verifying the relaying
+transaction. `messageHash` SHOULD be `indexed` to allow for easy filtering of logs.
+
+A cross chain message is defined in the following format:
+
+```solidity
+struct CrossChainMessage {
+    uint256 nonce;
+    address sender;
+    address target;
+    uint256 value;
+    uint256 timestamp;
+    uint256 chainid;
+    bytes data;
+}
+```
+
+Note that there is no `gasLimit` to avoid complexity with EVM gas introspection.
+
+The `messageHash` is constructed like so:
+
+```solidity
+function hashCrossChainMessage(CrossChainMessage memory _msg) pure returns (bytes32) {
+    return keccak256(abi.encode(_msg.nonce, _msg.sender, _msg.target, _msg.value, _msg.timestamp, _msg.chainid, _msg.data));
+}
+```
+
+### CrossL2Inbox
+
+Address: `0x4200000000000000000000000000000000000022`
+
+The `CrossL2Inbox` is responsible for relaying a cross chain message on the remote chain.
+It is permissionless to finalize a cross chain message on behalf of any user. Certain protocol
+enforced invariants must be preserved to ensure safety of the protocol.
+
+#### Invariants
+
+- The timestamp of relaying message MUST be greater than or equal to the timestamp of the initiating message
+- The chain id of the initiating message MUST be in the interop set
+- The relaying message MUST be initiated by an externally owned account such that the top level EVM call frame enters the `CrossL2Inbox`
```

Comment:
Just wondering the rationale behind this. Is there some vulnerability in allowing an EOA to call a contract that relays a message? Does this mean EIP-4337 accounts/paymasters are barred from relaying messages?

Reply:
This is a known limitation of the design. There is no way to have static analysis on a transaction without an `onlyEOA` check unfortunately. Without this, it will be very expensive for the block builder as it will need to do a full simulation to get the information required to verify the existence of the initiating transaction. This introduces denial of service attacks for the sequencer. App level builders will figure out other types of paymaster systems for this